compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,132 @@
 object FirMemberPropertyChecker : FirRegularClassChecker() {
     override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        for (member in declaration.declarations) {
-            if (member is FirProperty) {
-                checkProperty(declaration, member, context, reporter)
+        val memberProperties = declaration.declarations.filterIsInstance<FirProperty>()
+
+        // A property is known to be initialized only if it is initialized
+        //   1) with its own initializing expression;
+        //   2) at every class constructor;
+        //   3) at any of class's anonymous initializers; or
+        //   4) at other property's initializing expression
+
+        // 2) Property can be initialized at constructors. Since it's unknown what constructor will be used, the property can be determined
+        // as initialized only if it is initialized at every constructor. The caveat is a constructor that calls another constructor, e.g.,
+        //   constructor() { x = ... }
+        //   constructor(...): this() { ... }  // x will be initialized via this() delegation
+
+        // 3) Property can be initialized at any of class's anonymous initializers (all of initializers will be executed), e.g.,
+        //   init { x = ... }
+        //   ...
+        //   init { y = ... }
+
+        // 4) Property can be initialized at other property's initializing expression too, e.g.,
+        //   val initX = inlineMe { x = ... } // where inlineMe returns the value of the last expression of the lambda
+
+        // Here, we accumulate initialized info of a property per constructor, anonymous initializer, or property initialization
+        val initializedInConstructor = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+        val initializedInInit = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+        val initializedInOtherProperty = mutableMapOf<FirProperty, Boolean>().withDefault { false }
+
+        // To handle the delegated constructor call, we need a cache from constructor to (analyzed) property init info.
+        val constructorToData = mutableMapOf<FirConstructor, PropertyInitializationInfo>()
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirProperty, Boolean>,
+            acc: (Boolean, Boolean) -> Boolean,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo =
+                if (delegatedConstructor != null && constructorToData.containsKey(delegatedConstructor)) {
+                    constructorToData[delegatedConstructor]
+                } else null
+
+            val data = PropertyInitializationInfoCollector(memberProperties.map { it.symbol }.toSet()).getData(graph)
+            val infoAtExitNode = data[graph.exitNode]?.get(NormalPath)
+
+            val info = when {
+                delegatedInfo == null && infoAtExitNode == null -> return
```

Comment:
Consider also short circuiting the case when infoAtExitNode is empty

Reply:
Good point! I tried to add short circuits if either (or both) is empty, and then found a regression, which indicates that early bail out is wrong here. So, depending on what `acc` we do, e.g., `and` logic, empty info is still meaningful, like, nothing is initialized in the given CFG, and for `and`, that means, targeting property _could_ be not initialized.

Rather, I changed default-ish assignment a little bit, and avoid too-early bail-outs instead. Thanks!